runtime/sema/check_composite_declaration.go

```diff
@@ -522,6 +522,29 @@ func (checker *Checker) declareCompositeMembersAndValue(
 				VariableKind:    ast.VariableKindConstant,
 			}
 		}
+
+		// Declare implicit type requirement conformances, if any,
+		// after nested types are declared, and
+		// after explicit conformances are declared.
+		//
+		// For each nested composite type, check if a conformance
+		// declares a nested composite type with the same identifier,
+		// in which case it is a type requirement,
+		// and this nested composite type implicitly conforms to it.
+
+		for nestedTypeIdentifier, nestedType := range compositeType.NestedTypes() {
+			nestedCompositeType, ok := nestedType.(*CompositeType)
+			if !ok {
+				continue
+			}
+
+			for _, compositeTypeConformance := range compositeType.ExplicitInterfaceConformances {
+				conformanceNestedTypes := compositeTypeConformance.NestedTypes()
+				if typeRequirement, ok := conformanceNestedTypes[nestedTypeIdentifier].(*CompositeType); ok {
```

Comment:
Could there be a case where the `conformanceNestedTypes` do not contain the a type identifier that matches `nestedTypeIdentifier`?

I assume it is it a pre-condition (not Cadence pre-condition heh) that the `compositeTypeConformance` nested types are a subset of the `compositeType` nested types.

Reply:
Yes, this is the case when the conformance does not have a type requirement, e.g.

```
contract interface CI {}

contract C: CI {
    resource R {}
}
```
 
Here, `declareCompositeMembersAndValue` is called for `C`. It has one nested type `R` (first for loop above), which is a composie. We then look in all explicit interface conformances (`C` specifies it conforms to `CI`, and check if it has a nested type with the same name `R` – it does not, so the inner-most if-statement / line below / `AddImplicitTypeRequirementConformance` isn't called: `R` doesn't implicitly conform to any type requirements.

However, if `CI` would have a type requirement, then `C.R` implicitly conforms to it / `CI.R`, i.e. we call `AddImplicitTypeRequirementConformance` on `C.R` with argument `CI.R`

```
contract interface CI {
    resource R {}
}
```